pkg/config/resource.go

```diff
@@ -104,6 +104,12 @@ type ExternalName struct {
 	// assigned by the provider, like AWS VPC where it gets vpc-21kn123 identifier
 	// and not let you name it.
 	DisableNameInitializer bool
+
+	// IdentifierFields are the fields that are used to construct external
+	// resource identifier. We need to know these fields no matter what the
+	// management policy is including the Observe Only, different from other
+	// (required) fields.
+	IdentifierFields []string
```

Comment:
> but there are even anonymous config.ExternalName instances which will not be covered.

Could you point to some examples? Can we configure them as part of rolling the feature out to the providers?

> If this is not properly configured (i.e., if an identifier parameter is not correctly marked as so), we would expect the Terraform validation to catch it, but this would still be a compromise in our API quality. 

To be clear, if an identifier parameter is not marked correctly, we will replace `required` annotation with CEL validation rule which marks it as required. So, the impact is:

Default (not Observe Only):
- 1.25+ => No difference, the user will get warned during resource creation.
- 1.24- => Resource will be created and we get an error after the first reconciliation

Observe Only:
- User may create the resource without providing an identifier field but should get an error after the first reconciliation.

> I'm a little bit worried that we introduce a new external name configuration field without a robust defaulting for it.

> Is there a way to extract this information regarding which Terraform configuration arguments are part of the Terraform ID from the native schema? What do you think?

I agree with your concerns and have already investigated ways to figure this out before introducing a new configuration; however, unfortunately, couldn't find one. I believe this is a similar problem to external name configuration. The information could be inferred from the import documentation, but there is no robust way of automating this without human intervention. So, the best I can find is to leverage that existing configuration to figure the individual parameters out. Happy to try if you have some other ideas.

Reply:
What I mean with anonymous `config.ExternalName` instances are those which we anonymously define instead of a reusable definition (such as functions like `config.TemplatedStringAsIdentifier` or variables like the `config.IdentifierFromProvider`). An example is [here](https://github.com/upbound/provider-gcp/blob/f062d96b20f18247a5d274dc85589ff15c8cb460/config/sql/config.go#L117). 

As we have discussed offline, we may start getting reports on cases where we cannot enforce required fields at the Kubernetes API level as you mentioned above and although the remedy is to supply a value for the required field at runtime, we will also want to do a manual configuration on `config.ExternalName.IdentifierFields` as a permanent fix. This looks like to be the best option unless we extend the discussion to XRM or related stuff, so I believe we are good to go.